tortoise/verifying.go

```diff
@@ -149,39 +170,43 @@ func (v *verifying) verify(logger log.Log, lid types.LayerID) bool {
 		log.Stringer("local_threshold", v.localThreshold),
 		log.Stringer("global_threshold", v.globalThreshold),
 	)
-
-	// 0 - there is not enough weight to cross threshold.
-	// 1 - layer is verified and contextual validity is according to our local opinion.
 	if sign(margin.Cmp(v.globalThreshold)) == abstain {
-		logger.With().Info("candidate layer is not verified." +
-			" voting weight from good ballots is lower than the threshold")
+		logger.With().Debug("doesn't cross global threshold")
 		return false
 	}
-
-	// if there is any block with neutral local opinion - we can't verify the layer
-	// if it happens outside hdist - protocol will switch to full tortoise
-	for _, bid := range v.blocks[lid] {
-		vote, _ := getLocalVote(v.commonState, v.Config, lid, bid)
-		if vote == abstain {
-			logger.With().Info("candidate layer is not verified."+
-				" block is undecided according to local votes",
-				bid,
-			)
-			return false
-		}
-		v.validity[bid] = vote
+	if verifyLayer(
+		logger,
+		v.blocks[lid],
+		v.validity,
+		func(block blockInfo) sign {
+			decision, _ := getLocalVote(v.commonState, v.Config, lid, block.id)
+			if block.height > v.getReferenceHeight(lid) {
+				return abstain
+			}
+			return decision
+		},
+	) {
+		v.totalGoodWeight.Sub(v.goodWeight[lid])
+		return true
 	}
-
-	v.totalGoodWeight.Sub(v.goodWeight[lid])
-	logger.With().Info("candidate layer is verified")
-	return true
+	return false
 }
 
-func (v *verifying) countGoodBallots(logger log.Log, ballots []tortoiseBallot) (util.Weight, int) {
+func (v *verifying) countGoodBallots(logger log.Log, lid types.LayerID, ballots []tortoiseBallot) (util.Weight, int) {
 	sum := util.WeightFromUint64(0)
 	n := 0
 	for _, ballot := range ballots {
 		if ballot.weight.IsNil() {
+			logger.With().Debug("ballot weight is nil", ballot.id)
+			continue
+		}
+		// get height of the max votable block
+		if refheight := v.getReferenceHeight(lid.Sub(1)); refheight > ballot.height {
```

Comment:
i am still confused about this. in your reply to my earlier questions

> in general right, in this change median is used for everyone. since this is not a security thing, it feels like a mess if results won't be reproducible because of some difference between nodes that can't be fixed by downloading more data.

but the case i am thinking is:
minerH: miner height 100
refH: epoch medium height 110
lyrRefH: layer ref height 109
blockH: block A height 108
ballotH: ballot B height 120

in this case, can we consider ballot B a good ballot? theoretically miner cannot vote for a block A, but ballot B can. but it seems this implementation allows miner to vote for block A anyway?

Reply:
> but it seems this implementation allows miner to vote for block A anyway?

miner height is not relevant by itself. the only relevant height is a reference height. it could be equal to miner height. but i am using median height everywhere, so that tortoise will work consistently across the network. 

this is not about allowing miner to vote, this is about what votes will be counted. i am somewhat confused by this question

Reply:
in the example above you are not providing a ballot from that miner. if it had a ballot with height=100 - it won't be counted.

Reply:
> this is not about allowing miner to vote, this is about what votes will be counted

right. as you can see i am consistently confusing voting vs counting votes... 🙄